#include "connectionscreen.h"
#include "ui_connectionscreen.h"
#include "mainwindow.h"
#include <QDebug>
#include <QTimer>
#include <QCoreApplication>
#include <QEventLoop>
#include <QProcess>
#include <QListView>

ConnectionScreen::ConnectionScreen(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::ConnectionScreen),
    discoveryAgent(new QBluetoothDeviceDiscoveryAgent(this)),
    socket(nullptr),
    connectionCheckTimer(new QTimer(this))
{
    ui->setupUi(this);

    ui->lblConnection->setText("üî¥ Not Connected");

    connect(ui->btnScan, &QPushButton::clicked, this, &ConnectionScreen::updateBluetoothDevices);
    connect(ui->btnConnect, &QPushButton::clicked, this, &ConnectionScreen::connectToDevice);
    connect(ui->btnDisconnect, &QPushButton::clicked, this, &ConnectionScreen::disconnectDevice);
    connect(ui->btnClear, &QPushButton::clicked, this, &ConnectionScreen::clearLog);
    //connect(ui->btnRefresh, &QPushButton::clicked, this, &ConnectionScreen::refreshConnection);


    connect(discoveryAgent, &QBluetoothDeviceDiscoveryAgent::deviceDiscovered, this, [this](const QBluetoothDeviceInfo &device) {
        QString deviceName = device.name().trimmed();
        QString macAddress = device.address().toString();
        int rssi = device.rssi();
        qDebug() << "üì∂ Cihaz RSSI:" << rssi;

        if (deviceName.isEmpty()) {
            deviceName = macAddress;
        }

        /*if (rssi == 0) {
            qDebug() << "‚ö†Ô∏è Ge√ßersiz RSSI, cihaz atlandƒ±:" << device.address().toString();
            return; //
        }*/



        const auto deviceStr = QString("%1=%2").arg(deviceName).arg(macAddress);

        qDebug() << "Found Device:" << deviceName << "| MAC:" << macAddress;

        if (!devicess.contains(deviceStr)) {
            devicess[deviceStr] = macAddress;
            rssiValues[macAddress] = rssi;
            ui->comboBox->addItem(deviceStr);
        }
    });

    connect(connectionCheckTimer, &QTimer::timeout, this, [=]() {
        qDebug() << "Timer tetiklendi";
        if (socket)
            qDebug() << "Socket status : " << socket;
        if (socket)
            qDebug() << "Socket is open : " << socket->isOpen();
    });
    connectionCheckTimer->start(3000);

    QListView *comboListView = new QListView(ui->comboBox);
    comboListView->setMaximumHeight(500);
    comboListView->setFixedWidth(280);
    comboListView->setStyleSheet(
        "QComboBox::drop-down {"
        "   border: none; "
        "   background-color: #2E2E2E; "
        "   width: 20px; "
        "   padding: 0px; "
        "   margin: 0px; "
        "}"
        );
    ui->comboBox->setView(comboListView);
}

ConnectionScreen::~ConnectionScreen() {
    delete ui;
}

void ConnectionScreen::updateBluetoothDevices() {

    ui->comboBox->setUpdatesEnabled(false);
    ui->comboBox->clear();
    devicess.clear();
    ui->lblConnection->setText("üîç Scanning for devices...");
    ui->txtLog->append("üîç Scanning for devices...");

    if (discoveryAgent->isActive()) {
        discoveryAgent->stop();
    }

    discoveryAgent->start();
    ui->comboBox->setUpdatesEnabled(true);
}

void ConnectionScreen::connectToDevice() {
    QString selectedDevice = ui->comboBox->currentText();
    QString bluetoothVersion = "Unknown";
    bool connectedSuccessfully = false;


    if (socket && socket->isOpen()) {
        ui->txtLog->append("‚ö†Ô∏è Device is already connected: " + selectedDevice);
        ui->btnConnect->setEnabled(false);
        return;
    }

    if (selectedDevice.isEmpty() || !devicess.contains(selectedDevice)) {
        ui->txtLog->append("‚ö†Ô∏è Please select a device!");
        return;
    }

    QString macAddress = devicess[selectedDevice];
    int rssi = rssiValues.value(macAddress, -99);

    QBluetoothAddress bluetoothAddress(macAddress);
    QBluetoothUuid serviceUuid("0000110a-0000-1000-8000-00805f9b34fb");

    if (socket) {
        delete socket;
        socket = nullptr;
    }

    socket = new QBluetoothSocket(QBluetoothServiceInfo::RfcommProtocol, this);

    ui->lblConnection->setText("‚è≥ Connecting...");
    ui->lblConnection->setStyleSheet("color: yellow; font-weight: bold;");
    ui->txtLog->append("Connecting to device: " + selectedDevice);

    qDebug() << "üîπ [TEST] Trying to connect to Bluetooth device...";
    qDebug() << "   Selected Device: " << selectedDevice;
    qDebug() << "   MAC Address: " << macAddress;
    qDebug() << "   UUID: " << serviceUuid.toString();

    socket->connectToService(bluetoothAddress, serviceUuid);
    qDebug()<< "connecttoservice √ßaƒürƒ±ldƒ±";
    auto finalizeConnection = [=]() {
        if (socket && socket->isOpen()) {
            qDebug() << "‚úÖ Baƒülantƒ± ba≈üarƒ±lƒ±, emit deviceConnected";
            emit deviceConnected(selectedDevice, macAddress, rssiValues.value(macAddress, -99), bluetoothVersion);
            ui->lblConnection->setText("‚úÖ Connected: " + selectedDevice);
            ui->lblConnection->setStyleSheet("color: green; font-weight: bold; font-size: 18px;");
            ui->txtLog->append("‚úÖ Connected to device: " + selectedDevice);

        int rssi = rssiValues.value(macAddress, -99);
        double estimatedN = guessNFromRSSI(rssi);
        double distance = calculateDistance(-59, rssi, estimatedN);
        qDebug() << "üìè Estimated distance: " << distance << " metre";

        ui->txtLog->append("üì° RSSI: " + QString::number(rssi) + " dBm ");

        ui->txtLog->append("üìè Estimated distance: " + QString::number(distance, 'f', 2) + " m");
        //ui->txtLog->append("üìè Mesafe: " + QString::number(distance, 'f', 2) + " m");
        updateConnectionStatusLabel(true);
        }

        else {
            qDebug() << "‚ùå finalizeConnection: socket a√ßƒ±k deƒüil!";
        }

    };

//Timerƒ±n eklenme sebebi signal slotun yakalanmamasƒ±ndan dolayƒ± eklenmi≈ütir.
   /* QTimer::singleShot(4000, this, [=]() {
        if (socket && socket->isOpen()) {
            qDebug() << "‚úÖ GE√á gelen baƒülantƒ± algƒ±landƒ±.";
            emit deviceConnected(selectedDevice, macAddress, rssiValues.value(macAddress, -99), bluetoothVersion);
        } else {
            qDebug() << "‚ùå 4 saniye sonunda hala baƒülantƒ± yok.";
        }
    });


    connect(socket, &QBluetoothSocket::connected, this, [=]() {
        qDebug() << "üîπ [TEST] ";
        ui->lblConnection->setText("‚úÖ Connected: " + selectedDevice);
        ui->lblConnection->setStyleSheet("color: green; font-weight: bold; font-size: 18px;");
        ui->txtLog->append("‚úÖ Successfully connected to device: " + selectedDevice);

        /*int rssi = -99;
        if (rssiValues.contains(macAddress)) {
            rssi = rssiValues[macAddress];
        }
        qDebug() << "‚ö° Emit √ßaƒürƒ±lƒ±yor: deviceConnected";

    emit deviceConnected(selectedDevice, macAddress, rssi, bluetoothVersion);

   });

    QTimer::singleShot(3000, this, [=]() {
        if (socket->isOpen()) {
            qDebug() << "üîπ [TEST] Socket is still open, ensuring connection label update!";
            ui->lblConnection->setText("‚úÖ Connected: " + selectedDevice);
            ui->lblConnection->setStyleSheet("color: green; font-weight: bold; font-size: 18px;");
        }

    });
*/

    connect(socket, &QBluetoothSocket::connected, this, finalizeConnection);
    QTimer::singleShot(4000, this, finalizeConnection);
    connect(socket, &QBluetoothSocket::errorOccurred, this, [=](QBluetoothSocket::SocketError error) {
        ui->lblConnection->setText("‚ö†Ô∏è Connection Failed!");
        ui->lblConnection->setStyleSheet("color: red; font-weight: bold;");
        ui->txtLog->append("‚ö†Ô∏è Connection Error: " + socket->errorString());
        qDebug() << "‚ö†Ô∏è Error Code: " << error;
        qDebug() << "‚ö†Ô∏è Connection Error: " << socket->errorString();

    });
}

void ConnectionScreen::disconnectDevice() {
    if (!socket || !socket->isOpen()) {
        ui->txtLog->append("‚ö†Ô∏è No active connection.");
        ui->lblConnection->setText("üî¥ Not Connected");
        return;
    }

    ui->lblConnection->setText("‚è≥ Disconnecting...");
    ui->txtLog->append("Disconnecting from device...");
    qDebug() << "‚ö° Emit √ßaƒürƒ±lƒ±yor: devicedConnected";
   // emit deviceConnected(selectedDevice, macAddress, rssi, bluetoothVersion);
    if (socket->isOpen()) {
        qDebug() << "Disconnecting from Bluetooth service...";
        socket->disconnectFromService();
        socket->close();
        qDebug() << "Bluetooth connection closed.";
    }

    if (devicess.contains(ui->comboBox->currentText())) {
        QString macAddress = devicess[ui->comboBox->currentText()];
        QString command = "bluetoothctl disconnect " + macAddress;

        QProcess *process = new QProcess(this);
        connect(process, QOverload<QProcess::ProcessError>::of(&QProcess::errorOccurred), this, [=](QProcess::ProcessError error) {
            ui->txtLog->append("‚ö†Ô∏è Process Error: " + process->errorString());
            process->deleteLater();
        });
        connect(process, QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished), this, [=](int exitCode, QProcess::ExitStatus exitStatus) {
            if (exitStatus == QProcess::NormalExit && exitCode == 0) {
                ui->txtLog->append("Bluetoothctl disconnect command executed successfully.");
            } else {
                ui->txtLog->append("‚ö†Ô∏è Bluetoothctl disconnect command failed.");
            }
            process->deleteLater();
        });
        process->start("bluetoothctl", QStringList() << "disconnect" << macAddress);
    }

    if (socket) {
        delete socket;
        socket = nullptr;
    }

    ui->lblConnection->setText("‚ùå Disconnected");
    ui->txtLog->append("Bluetooth connection has been fully terminated.");
    ui->btnConnect->setEnabled(true);
    updateConnectionStatusLabel(false);

    //qDebug() << "üì§ emit bluetoothDisconnected() g√∂nderiliyor!";
    //emit bluetoothDisconnected();
}
double ConnectionScreen::calculateDistance(int measuredPower, int rssi, double N)
{
    // Distance = 10 ^ ((Measured Power - RSSI) / (10 * N))
    double exponent = (measuredPower - rssi) / (10.0 * N);
    double distance = pow(10.0, exponent);
    return distance;
}

double ConnectionScreen::guessNFromRSSI(int rssi) {
    if (rssi >= -50) return 1.5;
    if (rssi >= -60) return 2.0;
    if (rssi >= -70) return 2.5;
    if (rssi >= -80) return 3.0;
    return 3.5;
}

void ConnectionScreen::clearLog() {
    ui->txtLog->clear();
}

void ConnectionScreen::setConnectionLabelText(const QString& text, const QString& color) {
    ui->lblConnection->setText(text);
    ui->lblConnection->setStyleSheet("color: " + color + "; font-weight: bold;");
}
void ConnectionScreen::updateConnectionStatusLabel(bool connected) {
    if (connected) {
        ui->lblConnection->setText("‚úÖ Connected");
        ui->lblConnection->setStyleSheet("color: green; font-weight: bold;");
    } else {
        ui->lblConnection->setText("üî¥ Not Connected");
        ui->lblConnection->setStyleSheet("color: red; font-weight: bold;");
    }
}




